packages/bitcore-wallet-service/src/lib/server.ts

```diff
@@ -8897,6 +8905,633 @@ export class WalletService {
+            }
+            device = d;
+            next();
+          });
+        },
+        next => {
+          if (!device || !device.isValid()) {
+            device = LogDevice.create({
+              deviceId,
+              location
+            });
+          } else {
+            device.touch();
+            if (location) device.location = location;
+            if (token) device.token = token;
+            if (attendance) device.attendance();
+          }
+          next();
+        },
+        next => {
+          this.storage.updateLogDevice(device, next);
+        }
+      ],
+      err => {
+        if (err) {
+          return cb(err);
+        }
+        if (!device) {
+          return cb(new Error('Could not get current device for this deviceId'));
+        }
+
+        return cb(null, device);
+      }
+    );
+  }
+
+  /**
+   * Get Appreaciation by deviceId || ALL
+   * @param {Object} opts - Object fetch Appreciation.
+   * @param {String} deviceId - OPTIONAL. Fetch Appreciation by deviceId. NOT Fetch Appreciation all in DB.
+   * @returns {Array} - Return List Appreciation suitable condition.
+   */
+  getAllAppreciation(deviceId, cb) {
+    const opts = {
+      deviceId
+    };
+    this.storage.fetchAllAppreciation(opts, (err, listAppreciation) => {
+      if (err) return cb(err);
+      return cb(null, listAppreciation);
+    });
+  }
+
+  /**
+   * Update Appreaciation claimed
+   * @param {Object} opts - Object fetch Appreciation.
+   * @param {String} deviceId - DeviceId
+   * @param {String} claimCode - DeviceId
+   * @param {String} dateClaim - DeviceId
+   * @returns {Array} - Return List Appreciation suitable condition.
+   */
+  updateAppreciationClaim(opts, cb) {
+    const claimCode = opts?.claimCode;
+    const deviceId = opts?.deviceId;
+    const dateClaim = opts?.dateClaim;
+
+    let appreciation;
+
+    async.series(
+      [
+        next => {
+          this.storage.getAppreciationById({ deviceId, claimCode }, (err, appre) => {
+            if (err) {
+              return next(err);
+            }
+            appreciation = appre;
+            next();
+          });
+        },
+        next => {
+          if (appreciation) {
+            appreciation.deviceId = deviceId;
+            appreciation.dateClaim = dateClaim;
+            appreciation.status = !!dateClaim;
+            this.storage.updateAppreciation(appreciation, next);
+          }
+        }
+      ],
+      err => {
+        if (err) return cb(err);
+        if (!appreciation) {
+          return cb(new Error('Could not get current device for this deviceId'));
+        }
+
+        return cb(null, appreciation);
+      }
+    );
+  }
+
+  /**
+   * Apply Appreaciation for Device
+   * @param {String} deviceId - The token representing the app/device.
+   * @returns {Object} - Return Appreciation.
+   */
+  applyAppreciationForDevice(device, cb) {
+    let deviceId = device.deviceId;
+    let appreciation;
```

Comment:
I'm not really sure, but it seems we're able to not use the closure "appreciation" here by:
* In the first function of series, use next(null, appre);
* The params for second function is (appre, next), the result of first function is the input of the second

We can apply the same for other cases.

Reply:
I will use async waterfall instead of async series. It seems clear